src/com/facebook/buck/shell/DefaultWorkerTool.java

```diff
@@ -85,7 +103,47 @@ public int getMaxWorkers() {
   }
 
   @Override
+  public boolean isPersistent() {
+    return isPersistent;
+  }
+
+  @Override
   public ImmutableSortedSet<BuildRule> getRuntimeDeps() {
     return getDeps();
   }
+
+  @Override
+  public HashCode getInstanceKey() {
+    return buildOutputInitializer.getBuildOutput().getRuleKey().getHashCode();
+  }
+
+  @Override
+  public Data initializeFromDisk(OnDiskBuildInfo onDiskBuildInfo) throws IOException {
+    Optional<RuleKey> ruleKey =
+        onDiskBuildInfo.getRuleKey(BuildInfo.METADATA_KEY_FOR_RULE_KEY);
```

Comment:
Why does the worker tool need to read its rule key from the disk? Cannot it use a `RuleKeyBuilderFactory`? I would prefer that over relying on mutable external state.